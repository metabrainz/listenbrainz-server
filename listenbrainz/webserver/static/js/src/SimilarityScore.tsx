import * as React from "react";

export type SimilarityScoreProps = {
  type: "regular" | "compact";
  similarityScore: number;
  user?: ListenBrainzUser;
};

const getclassName = (similarityScore: number): string => {
  let className = "";
  if (similarityScore <= 0.3) {
    className = "red";
  } else if (similarityScore <= 0.7) {
    className = "orange";
  } else {
    className = "purple";
  }
  return className;
};

const SimilarityScore = (props: SimilarityScoreProps) => {
  const { similarityScore, user, type } = props;
  const className = getclassName(similarityScore);

  return (
    <div className={`similarity-score ${type}`}>
      <div
        className="progress"
        aria-label="Similarity Score"
        role="progressbar"
        aria-valuemin={0}
        aria-valuemax={100}
        aria-valuenow={similarityScore * 100}
        tabIndex={0}
      >
        <div
          className={`progress-bar ${className}`}
          style={{
            width: `${similarityScore * 100}%`,
          }}
        />
      </div>
      {type === "regular" ? (
        <p className="text-muted">
          Your compatibility with {user?.name} is {similarityScore * 10}/10
        </p>
      ) : (
        <p className="small text-muted">{similarityScore * 10}/10</p>
      )}
    </div>
  );
};

export default SimilarityScore;
