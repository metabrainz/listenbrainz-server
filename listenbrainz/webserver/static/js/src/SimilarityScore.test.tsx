import * as React from "react";
import { mount } from "enzyme";

import SimilarityScore, { SimilarityScoreProps } from "./SimilarityScore";

const props: SimilarityScoreProps = {
  similarityScore: 0.2,
  user: { auth_token: "baz", name: "test" },
  type: "regular",
};

describe("SimilarityScore", () => {
  it("renders correctly for type = 'regular'", () => {
    const wrapper = mount<SimilarityScoreProps>(<SimilarityScore {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly for type = 'compact'", () => {
    const wrapper = mount<SimilarityScoreProps>(
      <SimilarityScore {...{ ...props, type: "compact" }} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("updates the class name based on similiarty score", () => {
    /* sets class red for score 0.2 */
    let wrapper = mount<SimilarityScoreProps>(<SimilarityScore {...props} />);
    expect(wrapper.find(".progress").childAt(0).hasClass("red")).toEqual(true);

    /* sets class orange for score 0.5 */
    wrapper = mount<SimilarityScoreProps>(
      <SimilarityScore {...{ ...props, similarityScore: 0.5 }} />
    );
    expect(wrapper.find(".progress").childAt(0).hasClass("orange")).toEqual(
      true
    );

    /* sets class purple for score 0.9 */
    wrapper = mount<SimilarityScoreProps>(
      <SimilarityScore {...{ ...props, similarityScore: 0.9 }} />
    );
    expect(wrapper.find(".progress").childAt(0).hasClass("purple")).toEqual(
      true
    );
  });
});
