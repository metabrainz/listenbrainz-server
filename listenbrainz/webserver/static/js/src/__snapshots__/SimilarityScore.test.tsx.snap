// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimilarityScore renders correctly for type = 'compact' 1`] = `
<SimilarityScore
  similarityScore={0.2}
  type="compact"
  user={
    Object {
      "auth_token": "baz",
      "name": "test",
    }
  }
>
  <div
    className="similarity-score compact"
  >
    <div
      aria-label="Similarity Score"
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={20}
      className="progress"
      role="progressbar"
      tabIndex={0}
    >
      <div
        className="progress-bar red"
        style={
          Object {
            "width": "20%",
          }
        }
      />
    </div>
    <p
      className="small text-muted"
    >
      2
      /10
    </p>
  </div>
</SimilarityScore>
`;

exports[`SimilarityScore renders correctly for type = 'regular' 1`] = `
<SimilarityScore
  similarityScore={0.2}
  type="regular"
  user={
    Object {
      "auth_token": "baz",
      "name": "test",
    }
  }
>
  <div
    className="similarity-score regular"
  >
    <div
      aria-label="Similarity Score"
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={20}
      className="progress"
      role="progressbar"
      tabIndex={0}
    >
      <div
        className="progress-bar red"
        style={
          Object {
            "width": "20%",
          }
        }
      />
    </div>
    <p
      className="text-muted"
    >
      Your compatibility with 
      test
       is 
      2
      /10
    </p>
  </div>
</SimilarityScore>
`;
